### ISSUE TITLE
Bazel @HEAD frequently kills the server process

### ISSUE THREAD

**Issue by lberki** [bazelbuild/bazel#20276]
### Description of the bug:

When running Bazel in a loop, it occasionally kills the server for no good reason, on about every second invocation, with the error message `Killed non-responsive server process (pid=<number>)`

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```
touch WORKSPACE
cat > BUILD <<'EOF'
genrule(
    name = "gg",
    outs = ["go"],
    cmd = "sleep 1; echo O > $@",
)
EOF
for i in $(seq 1 20); do bazel build //:gg; done
```

Looping 20 times is usually enough to reproduce the issue frequently.

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

development version

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

From HEAD, commithash 08407743e903627d7e64c669f27b3f2b81316e33.

This issue doesn't seem to happen with 7.0.0rc4.

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
See the commithash above.
```

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

Somewhere between 7.0.0rc4 and 08407743e903627d7e64c669f27b3f2b81316e33; I haven't done the bisection.

**Comment by lberki** [comment]
Bisection says that the culprit is unfortunately 51bddee96bc357d8131efba606420227c7a8ebc5 , an Abseil version bump. I'm pretty certain: after this change, I got 2/20 restarts, before, 0/50 .

**Comment by lberki** [comment]
well that's alarming. I checked again and the confirmed the result of the bisection: 51bddee contains the bug, 39084edb2718865b6648274240ea94a29e62cdf7 does not and 7.0.0rc4 doesn't either (50 runs each, 51bddee fails very frequently so I'm pretty sure 39084edb is not a false negative)

Maybe we started using some Abseil functionality slightly differently between HEAD and 7.0.0 that happens to break with the upgrade?

**Comment by meteorcloudy** [comment]
@lberki How did you build the binary at 51bddee96bc357d8131efba606420227c7a8ebc5?

I actually couldn't reproduce this issue with Bazelisk with `USE_BAZEL_VERSION=51bddee96bc357d8131efba606420227c7a8ebc5`

**Comment by lberki** [comment]
Wat. `git checkout 51bddee`, then use the resulting Bazel binary.  I put the two binaries I have at `/google/data/ro/users/lb/lberki/bazel.restart/*` (commithashes in their names).

The issue readily reproduces on my workstation (asciinema recording on request :) )

**Comment by lberki** [comment]
The difference is that Bazelisk uses `-c opt` builds. Without `-c opt`, the issue reproduces for me at 51bddee, with it, it doesn't. More alarmingly, the issue also reproduces when I build Bazel from `release-7.0.0pre4` (8c4b1c908e5b520126a2480b0855c0c1e58d1ae0) without `-c opt`.

**Comment by lberki** [comment]
The good news is that this is reproducible also with `bazel version`. The bad news is that it becomes much less frequent with `GRPC_TRACE=all GRPC_VERBOSITY=INFO` and not reproducible with `GRPC_TRACE=all GRPC_VERBOSITY=DEBUG`.

I teased these lines out from the gRPC logs:

```
I1121 14:04:48.250881833 4115421 retry_filter.cc:896]        chand=0x564e161284f0 calld=0x564e16111070 attempt=0x564e160f0cc0: adding batch (start replayable pending batch on call attempt):  SEND_INITIAL_METADATA{:path: /command_server.CommandServer/Ping, grpc-timeout: 30S} SEND_MESSAGE:flags=0x00000000:len=33 SEND_TRAILING_METADATA{} RECV_INITIAL_METADATA RECV_MESSAGE RECV_TRAILING_METADATA
I1121 14:04:48.250898079 4115421 retry_filter.cc:1113]       chand=0x564e161284f0 calld=0x564e16111070 attempt=0x564e160f0cc0: starting 1 retriable batches on lb_call=0x564e16111900
I1121 14:04:48.250907137 4115421 call_combiner.h:172]        CallCombinerClosureList executing closure while already holding call_combiner 0x564e1610ceb0: closure=0x564e16111840 error=OK reason=start replayable pending bat
ch on call attempt
I1121 14:04:48.250942134 4115421 handshaker.cc:93]           handshake_manager 0x564e16119eb0: error=UNKNOWN:Invalid argument {file:"external/grpc~1.48.1.bcr.1/src/core/lib/iomgr/tcp_client_posix.cc", file_line:352, created_time:"2023-11-21T14:04:48.250314978+00:00", errno:22, os_error:"Invalid argument", syscall:"connect", target_address:"ipv6:%5B::1%5D:35397"} shutdown=0 index=1, args={endpoint=(nil), args=(nil) {size=0: {}}, read_buffer=(nil) (length=0), exit_early=0}
I1121 14:04:48.250972032 4115421 handshaker.cc:126]          handshake_manager 0x564e16119eb0: handshaking complete -- scheduling on_handshake_done with error=UNKNOWN:Invalid argument {file:"external/grpc~1.48.1.bcr.1/src/core/lib/iomgr/tcp_client_posix.cc", file_line:352, created_time:"2023-11-21T14:04:48.250314978+00:00", errno:22, os_error:"Invalid argument", syscall:"connect", target_address:"ipv6:%5B::1%5D:35397"}
```

Next step is `strace`, I guess, to see what exact system call fails and how it's different from the case when it doesn't fail?

**Comment by keertk** [comment]
@bazel-io fork 7.0.0

**Comment by lberki** [comment]
After pondering this a bit, I propose that we treat this as a release blocker until we find out what's going on. Given that  this bug manifests itself without `-c opt`, but not with it, I find it difficult to come up with a theory that makes it benign.

**Comment by lberki** [comment]
wdyt @meteorcloudy @meisterT ?

**Comment by meteorcloudy** [comment]
I can try to check if newer abseil-cpp version fixes it.

**Comment by meteorcloudy** [comment]
@meisterT Thanks, I can confirm `20230125.1` also fixes https://github.com/bazelbuild/bazel/issues/20076. I can send a change to update abseil to this version.

**Comment by benjaminp** [comment]
This is a gRPC bug: https://github.com/grpc/grpc/pull/35064

**Comment by lberki** [comment]
@benjaminp impressive sleuthing!

My guess is that you're right and this is in fact a gRPC bug and the reason why the Abseil update tickled it is that https://github.com/abseil/abseil-cpp/commit/3a46229c3cdd945ccc8bae1458148115ff8f88e7 added a `prctl()` call in some low-level allocation code that gets called somehow in between (dunno, e.g. by `grpc_sockaddr_to_uri()`). That commit was pushed in 2023 April, which is conveniently within the 2022 Jun 23 - 2023 Aug 2 time span covered by the Abseil update (it's a long time, though, so it's not a smoking gun)

The only part that's missing is explaining why this bug does not reproduce with `-c opt`. I haven't found a suspicious `#ifdef` but that doesn't mean that there isn't any given that it's low level memory allocation code, it's not surprising that it gets called at different times depending on how much the compiler optimizes things.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
